### PR TITLE
Add OIDC Provider support to TFVP

### PR DESCRIPTION
**Pull request by vinay-gopalan**
This PR adds OIDC Scope and Assignment resources to TFVP.

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccIdentityOIDCScope'
=== RUN   TestAccIdentityOIDCScope
--- PASS: TestAccIdentityOIDCScope (3.69s)
PASS

```

### CODE REVIEW

File: vault/resource_identity_oidc_assignment.go [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+
+import (
+	"fmt"
+	"log"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/vault/api"
+)
+
+const identityOIDCAssignmentPathTemplate = "identity/oidc/assignment"
```
Comment (benashz) [inline vault/resource_identity_oidc_assignment.go:11]: nit: this feels more like a pure path since it doesn't contain any interpolation variables, may want to consider renaming it.

File: vault/resource_identity_oidc_assignment_test.go [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+	name := acctest.RandomWithPrefix("test-scope")
+	resourceName := "vault_identity_oidc_assignment.test"
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:     func() { testutil.TestAccPreCheck(t) },
+		Providers:    testProviders,
+		CheckDestroy: testAccCheckIdentityEntityDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccIdentityOIDCAssignmentConfig_basic(name),
```
Comment (benashz) [inline vault/resource_identity_oidc_assignment_test.go:23]: Might be a good idea to add a test where `goup_ids` and `entity_ids` are empty.

File: vault/resource_identity_oidc_scope.go [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+	log.Printf("[DEBUG] Read OIDC Scope for %s", path)
+	if resp == nil {
+		log.Printf("[WARN] OIDC Scope %s not found, removing from state", path)
+		d.SetId("")
+		return nil
+	}
+
+	for _, k := range []string{"template", "description"} {
+		if err := d.Set(k, resp.Data[k]); err != nil {
+			return fmt.Errorf("error setting state key \"%s\" on OIDC Scope %s, err=%w", k, path, err)
```
Comment (benashz) [inline vault/resource_identity_oidc_scope.go:97]:
```suggestion
			return fmt.Errorf("error setting state key %q on OIDC Scope %q, err=%w", k, path, err)
```

**Review by benashz — COMMENTED** [review]
Initial pass, looking real good!

File: vault/resource_identity_oidc_assignment.go [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:        schema.TypeString,
+				ForceNew:    true,
+				Description: "The name of the assignment.",
+				Required:    true,
+			},
+			"entity_ids": {
+				Type:        schema.TypeList,
```
[the remark this answers is not in the document]
Comment (vinay-gopalan) [reply vault/resource_identity_oidc_assignment.go]: Fixed in 002b0ae

File: vault/resource_identity_oidc_assignment.go [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+				Description: "A list of Vault entity IDs.",
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+				Optional: true,
+			},
+			"group_ids": {
```
[the remark this answers is not in the document]
Comment (vinay-gopalan) [reply vault/resource_identity_oidc_assignment.go:35]: Fixed in 002b0ae

File: vault/resource_identity_oidc_assignment.go [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+func getOIDCAssignmentPath(name string) string {
+	return fmt.Sprintf("%s/%s", identityOIDCAssignmentPathTemplate, name)
+}
+
+func identityOIDCAssignmentCreateUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*api.Client)
+	name := d.Get("name").(string)
+	path := getOIDCAssignmentPath(name)
+
+	data := identityOIDCAssignmentRequestData(d)
```
[the remark this answers is not in the document]
Comment (vinay-gopalan) [reply vault/resource_identity_oidc_assignment.go]: Fixed in 002b0ae

File: vault/resource_identity_oidc_scope.go [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:        schema.TypeString,
+				ForceNew:    true,
+				Description: "The name of the scope. The openid scope name is reserved.",
+				Required:    true,
+			},
+			"template": {
+				Type:        schema.TypeString,
```
Comment (vinay-gopalan) [inline vault/resource_identity_oidc_scope.go:28]: @benashz working on making this into a map and encoding it to a string when writing it to Vault. What would the conversion look like? Since `EncodeToString` expects a `[]byte`, what's the cleanest way to convert a TypeMap to a `[]byte`? The following doesn't exactly work since you can't convert a map to a byte array

```
if v, ok := d.GetOk(k); ok {
	if k == "template" {
		data[k] = base64.StdEncoding.EncodeToString([]byte(v.(map[string]interface{})))
	} else {
		data[k] = v
	}
}
```

Comment (benashz) [reply vault/resource_identity_oidc_scope.go:28]: It looks like template is `TypeString`, which looks good. The map would be in the calling Terraform code like:

```hcl
template = jsonencode(
   foo = "bar"
)
```

The provider would take that as a string and pass it to vault base64 encoded.

We will probably want to add a `ValidateFunc` to ensure that the the JSON syntax is valid.

File: vault/resource_identity_oidc_assignment.go [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+	log.Printf("[DEBUG] Read OIDC Assignment for %s", path)
+	if resp == nil {
+		log.Printf("[WARN] OIDC Assignment %s not found, removing from state", path)
+		d.SetId("")
+		return nil
+	}
+
+	for _, k := range []string{"entity_ids", "group_ids"} {
+		if err := d.Set(k, resp.Data[k]); err != nil {
+			return fmt.Errorf("error setting state key \"%s\" on OIDC Assignment %s, err=%w", k, path, err)
```
[the remark this answers is not in the document]
Comment (vinay-gopalan) [reply vault/resource_identity_oidc_assignment.go]: Fixed

File: vault/resource_identity_oidc_assignment.go (hunk 1 above)
Comment (vinay-gopalan) [reply vault/resource_identity_oidc_assignment.go]: Updated to `Prefix` instead of `Template`

File: vault/resource_identity_oidc_assignment_test.go (hunk 2 above)
Comment (vinay-gopalan) [reply vault/resource_identity_oidc_assignment_test.go:33]: Updated with new empty case test

File: vault/resource_identity_oidc_scope.go [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+				Description: "The template string for the scope. This may be provided as escaped JSON or base64 encoded JSON.",
+				Optional:    true,
+			},
+			"description": {
+				Type:        schema.TypeString,
+				Description: "A description of the scope.",
```
[the remark this answers is not in the document]
Comment (vinay-gopalan) [reply vault/resource_identity_oidc_scope.go]: Fixed

File: vault/resource_identity_oidc_client.go [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+				Type:        schema.TypeString,
+				ForceNew:    true,
+				Description: "The name of the client.",
+				Required:    true,
+			},
+			"key": {
+				Type:        schema.TypeString,
+				ForceNew:    true,
+				Description: "A reference to a named key resource in Vault. This cannot be modified after creation.",
+				Required:    true,
```
Comment (austingebauer) [inline vault/resource_identity_oidc_client.go:31]: The key is optional in Vault 1.10. The "default" key will be used if no key is supplied.

File: vault/resource_identity_oidc_client.go [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:        schema.TypeString,
+				ForceNew:    true,
+				Description: "The name of the client.",
+				Required:    true,
+			},
+			"key": {
+				Type:        schema.TypeString,
+				ForceNew:    true,
```
Comment (austingebauer) [inline vault/resource_identity_oidc_client.go:29]: Nice! I think ForceNew is used for immutable fields? (trying to learn TF here :))

File: website/docs/r/identity_oidc_client.html.md [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+* `redirect_uris` - (Optional) Redirection URI values used by the client. 
+  One of these values must exactly match the `redirect_uri` parameter value
+  used in each authentication request.
+
+* `assignments` - (Optional) A list of assignment resources associated with the client.
+
+* `id_token_ttl` - (Optional) The time-to-live for ID tokens obtained by the client. 
+  The value should be less than the `verification_ttl` on the key.
+
+* `access_token_ttl` - (Optional) The time-to-live for access tokens obtained by the client.
```
Comment (austingebauer) [inline website/docs/r/identity_oidc_client.html.md:65]: Similar to comment above, `client_type` can be added here as a field.

**Review by austingebauer — COMMENTED** [review]
Overall, this looks great! Thanks @vinay-gopalan. Left a few minor suggestions and new additions coming in 1.10.

File: vault/resource_identity_oidc_scope.go (hunk 7 above)
Comment (vinay-gopalan) [reply vault/resource_identity_oidc_scope.go]: Fixed in 4848385

File: vault/data_identity_oidc_client_creds.go [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+		Read: readOIDCClientCredsResource,
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: "The name of the client.",
+			},
+			"client_id": {
+				Type:        schema.TypeString,
+				Description: "A reference to a named key resource in Vault. This cannot be modified after creation.",
```
[the remark this answers is not in the document]
Comment (vinay-gopalan) [reply vault/data_identity_oidc_client_creds.go:22]: Whoops! Bad copy-paste 😅

File: vault/data_identity_oidc_client_creds.go [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+				Computed:    true,
+			},
+			"client_secret": {
+				Type:        schema.TypeString,
+				Description: "A reference to a named key resource in Vault. This cannot be modified after creation.",
+				Computed:    true,
```
[the remark this answers is not in the document]
Comment (vinay-gopalan) [reply vault/data_identity_oidc_client_creds.go:28]: That makes sense! On the TF side we can just mark this schema field as sensitive and the SDK will take care of the obfuscation for us

File: vault/resource_identity_oidc_client.go (hunk 11 above)
Comment (vinay-gopalan) [reply vault/resource_identity_oidc_client.go:29]: @austingebauer exactly! The TF SDK will let the user know that they are trying to change an immutable field, which would result in the destruction of the previously configured resource and creation of a new Client resource :)

**Review by benashz — COMMENTED** [review]
Added a few new comments/nits. The resources are looking great!

I think the only thing left are the data sources. I will make another pass when they are ready.

File: vault/data_identity_oidc_client_creds.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (austingebauer) [reply vault/data_identity_oidc_client_creds.go:28]: Thanks, both! Makes sense 👍

File: vault/resource_identity_oidc_client.go (hunk 10 above)
Comment (vinay-gopalan) [reply vault/resource_identity_oidc_client.go:31]: @austingebauer wanted to confirm — then is this field mutable and can it be reset? In that case we will not force the creation of a new resource if the named key changes.

Comment (vinay-gopalan) [reply vault/resource_identity_oidc_client.go:31]: That makes sense. Thanks for confirming! 👍🏼

File: vault/data_identity_oidc_client_creds.go [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+				Description: "The name of the client.",
+			},
+			"client_id": {
+				Type:        schema.TypeString,
+				Description: "A reference to a named key resource in Vault. This cannot be modified after creation.",
+				Computed:    true,
+			},
+			"client_secret": {
+				Type:        schema.TypeString,
+				Description: "A reference to a named key resource in Vault. This cannot be modified after creation.",
```
[the remark this answers is not in the document]
Comment (vinay-gopalan) [reply vault/data_identity_oidc_client_creds.go]: Fixed!

File: vault/resource_identity_oidc_client.go [hunk 16]
```diff
@@ -0,0 +1,156 @@
...
+			"client_id": {
+				Type:        schema.TypeString,
+				Description: "The Client ID computed by and returned from Vault.",
+				Computed:    true,
+			},
+			"client_secret": {
+				Type:        schema.TypeString,
+				Description: "The Client Secret computed by and returned from Vault.",
+				Computed:    true,
+			},
```
[the remark this answers is not in the document]
Comment (vinay-gopalan) [reply vault/resource_identity_oidc_client.go:81]: Added new field in 1e9f8a3

File: vault/resource_identity_oidc_client.go [hunk 17]
```diff
@@ -0,0 +1,156 @@
...
+				Type:        schema.TypeInt,
+				Description: "The time-to-live for access tokens obtained by the client.",
+				Optional:    true,
+			},
+			"client_id": {
+				Type:        schema.TypeString,
+				Description: "The Client ID computed by and returned from Vault.",
+				Computed:    true,
+			},
+			"client_secret": {
```
[the remark this answers is not in the document]
Comment (vinay-gopalan) [reply vault/resource_identity_oidc_client.go:69]: Fixed!

File: vault/resource_identity_oidc_client.go (hunk 10 above)
Comment (vinay-gopalan) [reply vault/resource_identity_oidc_client.go]: Fixed in 1e9f8a3

File: website/docs/r/identity_oidc_client.html.md (hunk 12 above)
Comment (vinay-gopalan) [reply website/docs/r/identity_oidc_client.html.md:65]: Added docs for new field in 1e9f8a3

File: vault/data_identity_oidc_openid_config.go [hunk 18]
```diff
@@ -0,0 +1,60 @@
...
+
+	resp, err := client.RawRequest(r)
+	if err != nil {
+		return fmt.Errorf("error performing GET at %s, err=%w", path, err)
+	}
+	if resp != nil {
+		defer resp.Body.Close()
+	}
+
+	secret, err := api.ParseSecret(resp.Body)
```
[the remark this answers is not in the document]
Comment (vinay-gopalan) [reply vault/data_identity_oidc_openid_config.go]: Fixed in 30da4de

File: vault/data_identity_oidc_openid_config.go [hunk 19]
```diff
@@ -0,0 +1,60 @@
...
+	r := client.NewRequest("GET", path)
+
+	resp, err := client.RawRequest(r)
+	if err != nil {
+		return fmt.Errorf("error performing GET at %s, err=%w", path, err)
+	}
+	if resp != nil {
+		defer resp.Body.Close()
+	}
+
```
[the remark this answers is not in the document]
Comment (vinay-gopalan) [reply vault/data_identity_oidc_openid_config.go:124]: Fixed in 30da4de

File: vault/data_identity_oidc_openid_config.go [hunk 20]
```diff
@@ -0,0 +1,60 @@
...
+func identityOIDCOpenIDConfigDataSource() *schema.Resource {
+	return &schema.Resource{
+		Read: readOIDCOpenIDConfigResource,
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: "The name of the provider.",
+			},
+			"openid_config": {
```
[the remark this answers is not in the document]
Comment (vinay-gopalan) [reply vault/data_identity_oidc_openid_config.go]: Added refined schema in 30da4de

File: vault/resource_identity_oidc_client.go [hunk 21]
```diff
@@ -0,0 +1,155 @@
...
+				Type:        schema.TypeString,
+				Description: "The Client Secret computed by and returned from Vault.",
```
[the remark this answers is not in the document]
Comment (vinay-gopalan) [reply vault/resource_identity_oidc_client.go]: Fixed

File: vault/resource_identity_oidc_provider.go [hunk 22]
```diff
@@ -0,0 +1,133 @@
...
+				Optional: true,
+			},
+		},
+	}
+}
+
+func identityOIDCProviderRequestData(d *schema.ResourceData) map[string]interface{} {
+	fields := []string{"issuer", "allowed_client_ids", "scopes_supported"}
+	data := map[string]interface{}{}
+
```
[the remark this answers is not in the document]
Comment (vinay-gopalan) [reply vault/resource_identity_oidc_provider.go]: Fixed

File: website/docs/d/identity_oidc_client_creds.html.md [hunk 23]
```diff
@@ -0,0 +1,69 @@
...
+  redirect_uris = [
+    "http://127.0.0.1:9200/v1/auth-methods/oidc:authenticate:callback",
+    "http://127.0.0.1:8251/callback",
+    "http://127.0.0.1:8080/callback"
+  ]
+  id_token_ttl     = 2400
+  access_token_ttl = 7200
+}
+
+data "vault_identity_oidc_client_creds" "test" {
```
[the remark this answers is not in the document]
Comment (vinay-gopalan) [reply website/docs/d/identity_oidc_client_creds.html.md]: Nope! Good catch, removed in latest commit

**Review by benashz — APPROVED** [review]
A few minor nits/suggestions, but overall LGTM!

File: vault/data_identity_oidc_openid_config.go [hunk 24]
```diff
@@ -39,21 +113,41 @@ func readOIDCOpenIDConfigResource(d *schema.ResourceData, meta interface{}) erro
 	if err != nil {
 		return fmt.Errorf("error performing GET at %s, err=%w", path, err)
 	}
+
+	if resp == nil {
+		return fmt.Errorf("expected a response body, got nil response")
+	}
+
 	if resp != nil {
```
[the remark this answers is not in the document]
Comment (vinay-gopalan) [reply vault/data_identity_oidc_openid_config.go]: Fixed in 540d559

File: vault/data_identity_oidc_openid_config_test.go [hunk 25]
```diff
@@ -0,0 +1,79 @@
...
+	"testing"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/acctest"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+
+	"github.com/hashicorp/terraform-provider-vault/testutil"
+)
+
+const (
+	issuer                = "http://127.0.0.1:8200/v1/identity/oidc/provider/%s"
```
[the remark this answers is not in the document]
Comment (vinay-gopalan) [reply vault/data_identity_oidc_openid_config_test.go]: Moved within test function's scope in 540d559

**Review by austingebauer — APPROVED** [review]
LGTM

File: vault/resource_identity_oidc_provider.go [hunk 26]
```diff
@@ -27,8 +28,60 @@ func identityOIDCProviderResource() *schema.Resource {
 			"issuer": {
 				Type: schema.TypeString,
 				Description: "Specifies what will be used as the 'scheme://host:port' component for the 'iss' claim of ID tokens." +
-					"If provided explicitly, it must point to a Vault instance that is network reachable by clients for ID token validation.",
+					"This value is computed using the issuer_host and https_enabled schema fields.",
 				Computed: true,
+				// TODO confirm if this is needed
+				Optional: true,
```
Comment (vinay-gopalan) [inline vault/resource_identity_oidc_provider.go:34]: @benashz I had to mark this field as `Optional` to make the `DiffSuppressFunc` usable, though I'm not sure if either marking it as Optional or the Diff Suppression is still necessary

File: vault/resource_identity_oidc_provider.go [hunk 27]
```diff
@@ -0,0 +1,211 @@
...
+				Type:        schema.TypeString,
+				ForceNew:    true,
+				Description: "The name of the provider.",
+				Required:    true,
+			},
+			"issuer": {
+				Type:     schema.TypeString,
+				ForceNew: true,
+				Description: "Specifies what will be used as the 'scheme://host:port' component for the 'iss' claim of ID tokens." +
+					"This value is computed using the issuer_host and https_enabled schema fields.",
```
Comment (benashz) [inline vault/resource_identity_oidc_provider.go:32]:
```suggestion
					"This value is computed using the issuer_host and https_enabled fields.",
```

File: vault/resource_identity_oidc_provider.go [hunk 28]
```diff
@@ -0,0 +1,211 @@
...
+				Computed: true,
+				// TODO confirm if this is needed
+				Optional: true,
+				DiffSuppressFunc: func(k, old, new string, d *schema.ResourceData) bool {
```
Comment (benashz) [inline vault/resource_identity_oidc_provider.go:36]: Probably don't need this since you changed the way that issuer is configured.

File: vault/resource_identity_oidc_provider.go [hunk 29]
```diff
@@ -0,0 +1,211 @@
...
+						oldURLParsed.Scheme == newURLParsed.Scheme {
+						return true
+					}
+
+					return false
+				},
+			},
+			"https_enabled": {
+				Type:        schema.TypeBool,
+				Description: "Specifies whether the issuer host is on a https server.",
```
Comment (benashz) [inline vault/resource_identity_oidc_provider.go:57]:
```suggestion
				Description: "Set to true if the issuer endpoint uses HTTPS",
```

**Review by benashz — APPROVED** [review]
A few minor nits/suggestions then 👍

File: vault/resource_identity_oidc_provider.go (hunk 27 above)
Comment (vinay-gopalan) [reply vault/resource_identity_oidc_provider.go]: Fixed in 68a0eb6

File: vault/resource_identity_oidc_provider.go (hunk 28 above)
Comment (vinay-gopalan) [reply vault/resource_identity_oidc_provider.go]: Removed in 68a0eb6

File: vault/resource_identity_oidc_provider.go (hunk 29 above)
Comment (vinay-gopalan) [reply vault/resource_identity_oidc_provider.go]: Fixed in 68a0eb6

File: vault/resource_identity_oidc_provider.go [hunk 30]
```diff
@@ -0,0 +1,169 @@
...
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:        schema.TypeString,
+				ForceNew:    true,
+				Description: "The name of the provider.",
+				Required:    true,
+			},
+			"issuer": {
+				Type:     schema.TypeString,
+				ForceNew: true,
```
Comment (austingebauer) [inline vault/resource_identity_oidc_provider.go:29]: Any reason for `ForceNew: true` here? This should be updatable without needing to delete/recreate the provider.

File: website/docs/d/identity_oidc_client_creds.html.md [hunk 31]
```diff
@@ -0,0 +1,62 @@
...
+the console output when Terraform runs, and may be included in plan files
+if secrets are interpolated into any resource attributes.
+Protect these artifacts accordingly. See
+[the main provider documentation](../index.html)
+for more details.
+
+## Example Usage
+
+```hcl
+resource "vault_identity_oidc_key" "key" {
```
Comment (austingebauer) [inline website/docs/d/identity_oidc_client_creds.html.md:24]: A key with name "default" is built-in to Vault 1.10, so this will be updating (not creating) that key. I think we can omit this resource block entirely and remove the reference in `vault_identity_oidc_client` below?

File: website/docs/r/identity_oidc_client.html.md [hunk 32]
```diff
@@ -0,0 +1,81 @@
...
+# vault\_identity\_oidc\_client
+
+Manages OIDC Clients in a Vault server. See the [Vault documentation](https://www.vaultproject.io/api-docs/secret/identity/oidc-provider#create-or-update-an-assignment)
+for more information.
+
+## Example Usage
+
+```hcl
+resource "vault_identity_oidc_key" "test" {
+  name               = "default"
```
Comment (austingebauer) [inline website/docs/r/identity_oidc_client.html.md:18]: Similar comment as above regarding "default" key.

File: website/docs/r/identity_oidc_client.html.md [hunk 33]
```diff
@@ -0,0 +1,81 @@
...
+}
+```
+
+## Argument Reference
+
+The following arguments are supported:
+
+* `name` - (Required) The name of the client.
+
+* `key` - (Required) A reference to a named key resource in Vault.
```
Comment (austingebauer) [inline website/docs/r/identity_oidc_client.html.md:53]: The key is optional in Vault 1.10.

File: website/docs/r/identity_oidc_provider.html.md [hunk 34]
```diff
@@ -0,0 +1,98 @@
...
+# vault\_identity\_oidc\_provider
+
+Manages OIDC Providers in a Vault server. See the [Vault documentation](https://www.vaultproject.io/api-docs/secret/identity/oidc-provider#create-or-update-an-assignment)
+for more information.
+
+## Example Usage
+
+```hcl
+resource "vault_identity_oidc_key" "test" {
+  name               = "default"
```
Comment (austingebauer) [inline website/docs/r/identity_oidc_provider.html.md:18]: Similar comment as above regarding "default" key.

File: website/docs/r/identity_oidc_provider.html.md [hunk 35]
```diff
@@ -0,0 +1,98 @@
...
+* `allowed_client_ids` - (Optional) The client IDs that are permitted to use the provider. 
+  If empty, no clients are allowed. If `*`, all clients are allowed.
+
+* `scopes_supported` - (Optional) The scopes available for requesting on the provider.
+
+## Attributes Reference
+
+In addition to the arguments above, the following attributes are exported:
+
+* `issuer` - (Required) Specifies what will be used as the `scheme://host:port`
```
Comment (austingebauer) [inline website/docs/r/identity_oidc_provider.html.md:87]: I wonder if `(Required)` makes sense for an attribute? I thought these were output only, but I might be misunderstanding.

File: vault/resource_identity_oidc_provider.go (hunk 30 above)
Comment (austingebauer) [reply vault/resource_identity_oidc_provider.go:29]: I guess this probably isn't an "input" for creating this resource. If I understand correctly, it's replaced by `https_enabled` and `issuer_host`.

**Review by austingebauer — APPROVED** [review]
LGTM otherwise. Thanks, @vinay-gopalan!

File: website/docs/r/identity_oidc_provider.html.md (hunk 34 above)
Comment (vinay-gopalan) [reply website/docs/r/identity_oidc_provider.html.md]: Thanks for catching these! Fixed in latest commit

File: website/docs/r/identity_oidc_provider.html.md (hunk 35 above)
Comment (vinay-gopalan) [reply website/docs/r/identity_oidc_provider.html.md]: Bad copypasta 😓  Fixed in latest commit!

File: website/docs/r/identity_oidc_client.html.md (hunk 33 above)
Comment (vinay-gopalan) [reply website/docs/r/identity_oidc_client.html.md]: Updated in latest commit

File: vault/resource_identity_oidc_provider.go (hunk 30 above)
Comment (vinay-gopalan) [reply vault/resource_identity_oidc_provider.go]: You are correct, Austin! Thanks for clarifying that `issuer` fields can be updated on OIDC Providers. Removed the `ForceNew` behavior from this field in the latest commit
